src/base/QXmppJingleData.cpp

```diff
@@ -2779,3 +2931,278 @@ bool QXmppJingleRtpHeaderExtensionProperty::isJingleRtpHeaderExtensionProperty(c
+}
+
+/// \cond
+void QXmppJingleMessageInitiationElement::parse(const QDomElement &element)
+{
+    d->type = stringToJmiElementType(element.nodeName());
+
+    if (d->type == Type::None) {
+        return;
+    }
+
+    d->id = element.attribute(QStringLiteral("id"));
+
+    // Type::Proceed and Type::Ringing don't need any parsing aside of the id.
+    switch (d->type) {
+    case Type::Propose: {
+        if (const auto &descriptionElement = element.firstChildElement("description"); !descriptionElement.isNull()) {
+            d->description = QXmppJingleDescription();
+            d->description->parse(descriptionElement);
+        }
+
+        break;
+    }
+    case Type::Reject:
+    case Type::Retract:
+        d->containsTieBreak = !element.firstChildElement("tie-break").isNull();
+
+        if (const auto &reasonElement = element.firstChildElement("reason"); !reasonElement.isNull()) {
+            d->reason = QXmppJingleReason();
+            d->reason->parse(reasonElement);
+        }
+
+        break;
+    case Type::Finish:
+        if (const auto &reasonElement = element.firstChildElement("reason"); !reasonElement.isNull()) {
+            d->reason = QXmppJingleReason();
+            d->reason->parse(reasonElement);
+        }
+
+        if (const auto &migratedToElement = element.firstChildElement("migrated"); !migratedToElement.isNull()) {
+            d->migratedTo = migratedToElement.attribute("to");
+        }
+
+        break;
+    default:
+        break;
+    }
+}
+
+void QXmppJingleMessageInitiationElement::toXml(QXmlStreamWriter *writer) const
+{
+    writer->writeStartElement(jmiElementTypeToString(d->type));
+    writer->writeDefaultNamespace(ns_jingle_message_initiation);
+
+    helperToXmlAddAttribute(writer, QStringLiteral("id"), d->id);
+
+    if (d->description) {
+        d->description->toXml(writer);
+    }
+
+    if (d->reason) {
+        d->reason->toXml(writer);
+    }
+
+    if (d->containsTieBreak) {
+        writer->writeEmptyElement(QStringLiteral("tie-break"));
+    }
+
+    if (!d->migratedTo.isEmpty()) {
+        writer->writeEmptyElement(QStringLiteral("migrated"));
+        helperToXmlAddAttribute(writer, QStringLiteral("to"), d->migratedTo);
+    }
+
+    writer->writeEndElement();
+}
+/// \endcond
+
+QXMPP_PRIVATE_DEFINE_RULE_OF_SIX(QXmppJingleMessageInitiationElement)
+
+///
+/// Returns true if passed QDomElement is a Jingle Message Initiation element
+///
+bool QXmppJingleMessageInitiationElement::isJingleMessageInitiationElement(const QDomElement &element)
+{
+    return (element.nodeName() == QStringLiteral("propose") || element.nodeName() == QStringLiteral("ringing") || element.nodeName() == QStringLiteral("proceed") || element.nodeName() == QStringLiteral("retract") || element.nodeName() == QStringLiteral("reject") || element.nodeName() == QStringLiteral("finish")) && element.hasAttribute(QStringLiteral("id")) && element.namespaceURI() == ns_jingle_message_initiation;
```

Comment:
* better use `stringToJmiElementType()` and check that the result is valid
* s/nodeName()/tagName()/